timestamp/token.go

```diff
@@ -0,0 +1,230 @@
+		if err != nil {
+			lastErr = err
+			continue
+		}
+		if _, err := signed.VerifySigner(ctx, &signerInfo, signingCertificate, opts); err != nil {
+			lastErr = err
+			continue
+		}
+		// RFC 3161 2.3: The corresponding certificate MUST contain only one instance of
+		// the extended key usage field extension.
+		if len(signingCertificate.ExtKeyUsage) == 1 && len(signingCertificate.UnknownExtKeyUsage) == 0 {
+			verifiedCerts = append(verifiedCerts, signingCertificate)
+		}
+	}
+	if len(verifiedCerts) == 0 {
+		var verificationErr cms.VerificationError
+		if errors.As(lastErr, &verificationErr) {
+			// TODO : add a debug logger here to log the verificationErr
+			return nil, errors.New("failed to verify timestamp token")
+		}
+		return nil, lastErr
+	}
+	return verifiedCerts, nil
+}
+
+// Info returns the timestamping information.
+func (t *SignedToken) Info() (*TSTInfo, error) {
+	var info TSTInfo
+	if _, err := asn1.Unmarshal(t.Content, &info); err != nil {
+		return nil, err
+	}
+	// RFC 3161 2.4.2 TSTInfo
+	if info.Version != 1 {
+		return nil, fmt.Errorf("timestamp token info version must be 1, but got %d", info.Version)
+	}
+	return &info, nil
+}
+
+// GetSigningCertificate gets the signing certificate identified by SignedToken
+// SignerInfo's SigningCertificate attribute. If the IssuerSerial
+// field of signing certificate is missing, use signerInfo's sid instead.
+// The identified signing certificate MUST match the hash in SigningCertificate.
+//
+// References: RFC 3161 2.4.1 & 2.4.2; RFC 5035 4
+func (t *SignedToken) GetSigningCertificate(ctx context.Context, signerInfo *cms.SignerInfo) (*x509.Certificate, error) {
+	var signingCertificateV2 SigningCertificateV2
+	if err := signerInfo.SignedAttributes.TryGet(oid.SigningCertificateV2, &signingCertificateV2); err != nil {
+		return nil, fmt.Errorf("failed to get SigningCertificateV2 from signed attributes: %w", err)
+	}
+	// get candidate signing certificate
+	var candidateSigningCert *x509.Certificate
+	signed := (*cms.ParsedSignedData)(t)
+	if signingCertificateV2.Certificates[0].IssuerSerial.Issuer.FullBytes != nil {
+		// use IssuerSerial from signingCertificateV2
+		var rawValue asn1.RawValue
+		_, err := asn1.Unmarshal(signingCertificateV2.Certificates[0].IssuerSerial.Issuer.Bytes, &rawValue)
+		if err != nil {
+			return nil, err
+		}
+		// RFC 3280 4.2.1.7
+		if rawValue.Tag != 4 {
+			return nil, fmt.Errorf("signing certificate IssuerSerial requires CHOICE tag 4, but got %d", rawValue.Tag)
+		}
+		var issuer asn1.RawValue
+		_, err = asn1.Unmarshal(rawValue.Bytes, &issuer)
+		if err != nil {
+			return nil, err
+		}
+		ref := cms.IssuerAndSerialNumber{
+			Issuer:       issuer,
+			SerialNumber: signingCertificateV2.Certificates[0].IssuerSerial.SerialNumber,
+		}
+		candidateSigningCert = signed.GetCertificate(ref)
+	} else {
+		// use sid as IssuerSerial
+		candidateSigningCert = signed.GetCertificate(signerInfo.SignerIdentifier)
```

Comment:
If we fallback to `sid`, will it violates rfc3161?

Reply:
We enforce the existence of `SigningCertificate` signed attribute as stated in RFC 3161, however, the `IssuerSerial` field under `ESSCertID` is indeed optional. In case of missing, we will fallback to sid to identify the cert. Once find the cert, we still compute its hash and compare with the value in `SigningCertificate`.